test/lib/jdk/test/lib/cds/CDSAppTester.java

```diff
@@ -0,0 +1,233 @@
+                                                   "-cp", classpath(runMode));
+        ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(cmdLine);
+        Process process = pb.start();
+        OutputAnalyzer output = CDSTestUtils.executeAndLog(process, "static");
+        listOutputFile(staticArchiveFile);
+        listOutputFile(staticArchiveFile + ".log");
+        checkExecutionHelper(output, runMode);
+        return output;
+    }
+
+    private OutputAnalyzer dumpDynamicArchive() throws Exception {
+        RunMode runMode = RunMode.DUMP_DYNAMIC;
+        String[] cmdLine = new String[0];
+        if (isDynamicWorkflow()) {
+          // "classic" dynamic archive
+          cmdLine = StringArrayUtils.concat(vmArgs(runMode), dynamicDumpLog(),
+                                            "-Xlog:cds",
+                                            "-XX:ArchiveClassesAtExit=" + dynamicArchiveFile,
+                                            "-cp", classpath(runMode));
+        }
+        cmdLine = StringArrayUtils.concat(cmdLine, appCommandLine(runMode));
+
+        ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(cmdLine);
+        Process process = pb.start();
+        OutputAnalyzer output = CDSTestUtils.executeAndLog(process, "dynamic");
+        listOutputFile(dynamicArchiveFile);
+        listOutputFile(dynamicArchiveFile + ".log");
+        checkExecutionHelper(output, runMode);
+        return output;
+    }
+
+    private OutputAnalyzer productionRun() throws Exception {
+        RunMode runMode = RunMode.PRODUCTION;
+        String[] cmdLine = StringArrayUtils.concat(vmArgs(runMode), productionRunLog(),
+                                                   "-cp", classpath(runMode));
+        if (isStaticWorkflow()) {
+            cmdLine = StringArrayUtils.concat(cmdLine, "-XX:SharedArchiveFile=" + staticArchiveFile);
+        } else if (isDynamicWorkflow()) {
+            cmdLine = StringArrayUtils.concat(cmdLine, "-XX:SharedArchiveFile=" + dynamicArchiveFile);
+        }
+
+        cmdLine = StringArrayUtils.concat(cmdLine, appCommandLine(runMode));
+
+        ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(cmdLine);
+        Process process = pb.start();
+        OutputAnalyzer output = CDSTestUtils.executeAndLog(process, "production");
+        listOutputFile(name() + ".production.log");
+        checkExecutionHelper(output, runMode);
+        return output;
+    }
+
+    public void run(String args[]) throws Exception {
+        if (args.length == 1) {
+            if (args[0].equals("STATIC")) {
+                runStaticWorkflow();
+                return;
+            }
+            if (args[0].equals("DYNAMIC")) {
+                runDynamicWorkflow();
+                return;
+            }
+        }
+
+        throw new RuntimeException("Must have exactly one command line argument: STATIC or DYNAMIC");
```

Comment:
Why not check the number of args at the beginning of the method?

Reply:
Fixed.